Build failure: srelay

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

1. check out latest `nixpkgs/master` or `nixpkgs/nixos-unstable`
3. `nix build .#srelay`

### Can Hydra reproduce this build failure?

Hydra is currently rebuilding this package.

### Link to Hydra build job

_No response_

### Relevant log output

```console
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/4yzhvs9sblbgllywshdrlfws4sb09jn3-srelay-0.4.8.tar.gz
source root is srelay-0.4.8
setting SOURCE_DATE_EPOCH to timestamp 1504255152 of file "srelay-0.4.8/srelay.h"
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
applying patch /nix/store/nl6yvxflcbv38rq5dyafcw15wyyxsi03-arm.patch
patching file configure
Hunk #1 succeeded at 2386 with fuzz 2 (offset 511 lines).
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Updating Autotools / GNU config script to a newer upstream version: ./config.sub
Updating Autotools / GNU config script to a newer upstream version: ./config.guess
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
patching script interpreter paths in ./configure
./configure: interpreter directive changed from "#! /bin/sh" to "/nix/store/2hjsch59amjs3nbgh7ahcfzm2bfwl8zi-bash-5.3p9/bin/sh"
configure flags: --prefix=/nix/store/7w7f7cr9xy5gpglpslm57l9w1cak17n7-srelay-0.4.8
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking target system type... x86_64-pc-linux-gnu
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /nix/store/8laf6k81j9ckylrigj3xsk76j69knhvl-gnugrep-3.12/bin/grep
checking for egrep... /nix/store/8laf6k81j9ckylrigj3xsk76j69knhvl-gnugrep-3.12/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking tcpd.h usability... no
checking tcpd.h presence... no
checking for tcpd.h... no
checking pthread.h usability... yes
checking pthread.h presence... yes
checking for pthread.h... yes
checking whether enabling thread... no
checking for crypt in -lcrypt... yes
checking for ANSI C header files... (cached) yes
checking for sys/wait.h that is POSIX.1 compatible... yes
checking fcntl.h usability... yes
checking fcntl.h presence... yes
checking for fcntl.h... yes
checking sys/time.h usability... yes
checking sys/time.h presence... yes
checking for sys/time.h... yes
checking limits.h usability... yes
checking limits.h presence... yes
checking for limits.h... yes
checking syslog.h usability... yes
checking syslog.h presence... yes
checking for syslog.h... yes
checking for unistd.h... (cached) yes
checking sys/resource.h usability... yes
checking sys/resource.h presence... yes
checking for sys/resource.h... yes
checking for socklen_t... yes
checking for u_int8_t... yes
checking for u_int16_t... yes
checking for u_int32_t... yes
checking for uint8_t... yes
checking for uint16_t... yes
checking for uint32_t... yes
checking whether time.h and sys/time.h may both be included... yes
checking for struct sockaddr_dl.sdl_family... no
checking for struct sockaddr.sa_len... no
checking return type of signal handlers... void
checking for select... yes
checking for socket... yes
checking for getaddrinfo... yes
checking for getnameinfo... yes
checking for freeaddrinfo... yes
configure: creating ./config.status
config.status: creating Makefile
config.status: creating config.h
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
build flags: SHELL=/nix/store/2hjsch59amjs3nbgh7ahcfzm2bfwl8zi-bash-5.3p9/bin/bash
gcc -I.  -DHAVE_CONFIG_H -DSYSCONFDIR=\"/nix/store/7w7f7cr9xy5gpglpslm57l9w1cak17n7-srelay-0.4.8/etc\" -O2 -Wall -Wno-unused-result -Wno-unused-but-set-variable -c init.c
init.c: In function 'serv_init':
init.c:128:44: warning: argument to 'sizeof' in 'strncpy' call is the same expression as the source; did you mean to use the size of the destination? [-Wsizeof-pointer-memaccess]
  128 |     strncpy(tmp_str_serv_sock, hbuf, sizeof(hbuf));
      |                                            ^
init.c:136:44: warning: argument to 'sizeof' in 'strncat' call is the same expression as the source; did you mean to use the size of the destination? [-Wsizeof-pointer-memaccess]
  136 |     strncat(tmp_str_serv_sock, sbuf, sizeof(sbuf));
      |                                            ^
gcc -I.  -DHAVE_CONFIG_H -DSYSCONFDIR=\"/nix/store/7w7f7cr9xy5gpglpslm57l9w1cak17n7-srelay-0.4.8/etc\" -O2 -Wall -Wno-unused-result -Wno-unused-but-set-variable -c readconf.c
readconf.c: In function 'str_to_addr':
readconf.c:359:7: warning: '__builtin_strncpy' output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
  359 |       strncpy((char *)dest->fqdn, addr, len);
      |       ^
readconf.c:357:16: note: length computed here
  357 |     if ((len = strlen(addr)) > 0 && len < 256) {
      |                ^~~~~~~~~~~~
readconf.c: In function 'getpasswd':
readconf.c:517:5: warning: '__builtin___strncpy_chk' output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
  517 |     strncpy(up->user, tok, len);
      |     ^
readconf.c:509:20: note: length computed here
  509 |     tok = p; len = strlen(tok);
      |                    ^~~~~~~~~~~
readconf.c:535:5: warning: '__builtin___strncpy_chk' output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
  535 |     strncpy(up->pass, tok, len);
      |     ^
readconf.c:528:20: note: length computed here
  528 |     tok = p; len = strlen(tok);
      |                    ^~~~~~~~~~~
gcc -I.  -DHAVE_CONFIG_H -DSYSCONFDIR=\"/nix/store/7w7f7cr9xy5gpglpslm57l9w1cak17n7-srelay-0.4.8/etc\" -O2 -Wall -Wno-unused-result -Wno-unused-but-set-variable -c util.c
util.c: In function 'setsignal':
util.c:301:16: error: assignment to '__sighandler_t' {aka 'void (*)(int)'} from incompatible pointer type 'sigfunc_t' {aka 'void (*)(void)'} [-Wincompatible-pointer-types]
  301 |   n.sa_handler = handler;
      |                ^
In file included from /nix/store/j8irrc0mpx029dw0rmadsjylg7h31ync-glibc-2.42-51-dev/include/sys/wait.h:36,
                 from util.c:37:
/nix/store/j8irrc0mpx029dw0rmadsjylg7h31ync-glibc-2.42-51-dev/include/signal.h:72:16: note: '__sighandler_t' declared here
   72 | typedef void (*__sighandler_t) (int);
      |                ^~~~~~~~~~~~~~
In file included from util.c:38:
srelay.h:369:27: note: 'sigfunc_t' declared here
  369 | typedef void            (*sigfunc_t)();
      |                           ^~~~~~~~~
util.c:308:12: error: returning 'void (*)(int)' from a function with incompatible return type 'sigfunc_t' {aka 'void (*)(void)'} [-Wincompatible-pointer-types]
  308 |     return SIG_ERR;
      |            ^~~~~~~
srelay.h:369:27: note: 'sigfunc_t' declared here
  369 | typedef void            (*sigfunc_t)();
      |                           ^~~~~~~~~
util.c:309:12: error: returning '__sighandler_t' {aka 'void (*)(int)'} from a function with incompatible return type 'sigfunc_t' {aka 'void (*)(void)'} [-Wincompatible-pointer-types]
  309 |   return o.sa_handler;
      |            ^~~~~~~~~~
srelay.h:369:27: note: 'sigfunc_t' declared here
  369 | typedef void            (*sigfunc_t)();
      |                           ^~~~~~~~~
/nix/store/j8irrc0mpx029dw0rmadsjylg7h31ync-glibc-2.42-51-dev/include/signal.h:72:16: note: '__sighandler_t' declared here
   72 | typedef void (*__sighandler_t) (int);
      |                ^~~~~~~~~~~~~~
make: *** [Makefile:25: util.o] Error 1
```

### Additional context

Hydra has reproduced this build failure on staging-next: https://hydra.nixos.org/build/323566165

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.18.13, NixOS, 26.05 (Yarara), 26.05.20260227.dd9b079`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Determinate Nix 3.16.3) 2.33.3`
 - channels(root): `""`
 - nixpkgs: `/nix/store/kzqqrhnhh9q3dnrj3x10cdihmj09yc8p-source`

### Notify maintainers



---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+label%3A%220.kind%3A+build+failure%22).
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/sr/srelay/package.nix
+++ b/pkgs/by-name/sr/srelay/package.nix
@@ -14,7 +14,10 @@ stdenv.mkDerivation (finalAttrs: {
     sha256 = "1sn6005aqyfvrlkm5445cyyaj6h6wfyskfncfmds55x34hfyxpvl";
   };
 
-  patches = [ ./arm.patch ];
+  patches = [
+    ./arm.patch
+    ./signal-handler.patch
+  ];
 
   buildInputs = [ libxcrypt ];
 

--- a/pkgs/by-name/sr/srelay/signal-handler.patch
+++ b/pkgs/by-name/sr/srelay/signal-handler.patch
@@ -1,0 +1,44 @@
+diff --git a/main.c b/main.c
+index 00d1cf3..3526cfd 100644
+--- a/main.c
++++ b/main.c
+@@ -675,7 +675,7 @@ int main(int argc, char **argv)
+     msg_out(warn, "cannot open pidfile %s", pidfile);
+   }
+ 
+-  setsignal(SIGHUP, reload);
++  setsignal(SIGHUP, do_sighup);
+   setsignal(SIGINT, SIG_IGN);
+   setsignal(SIGQUIT, SIG_IGN);
+   setsignal(SIGILL, SIG_IGN);
+@@ -690,7 +690,7 @@ int main(int argc, char **argv)
+   setsignal(SIGSYS, SIG_IGN);
+   setsignal(SIGPIPE, SIG_IGN);
+   setsignal(SIGALRM, SIG_IGN);
+-  setsignal(SIGTERM, cleanup);
++  setsignal(SIGTERM, do_sigterm);
+   setsignal(SIGUSR1, SIG_IGN);
+   setsignal(SIGUSR2, SIG_IGN);
+ #ifdef SIGPOLL
+@@ -737,7 +737,7 @@ int main(int argc, char **argv)
+     }
+   } else {
+ #endif
+-    setsignal(SIGCHLD, reapchild);
++    setsignal(SIGCHLD, do_sigchld);
+     setregid(-1, PROCGID);
+     setreuid(-1, PROCUID);
+     msg_out(norm, "Starting: MAX_CH(%d)", max_child);
+diff --git a/srelay.h b/srelay.h
+index 3e4d5e7..35010f0 100644
+--- a/srelay.h
++++ b/srelay.h
+@@ -366,7 +366,7 @@ struct user_pass {
+ };
+ 
+ #ifndef SIGFUNC_DEFINED
+-typedef void            (*sigfunc_t)();
++typedef void            (*sigfunc_t)(int);
+ #endif
+ 
+ #ifndef MAX

